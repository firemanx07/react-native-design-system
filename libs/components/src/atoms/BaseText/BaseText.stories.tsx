import { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { StorybookScreen } from '../../storybook';
import {
  BaseText as BaseTextBase,
  PropsType as BaseTextProps,
  TextVariant,
  TextWeight,
} from './BaseText';

const Story: ComponentMeta<typeof BaseTextBase> = {
  component: BaseTextBase,
  title: 'Atoms/Base Text',
  argTypes: {
    variant: {
      control: { type: 'radio' },
    },
    weight: {
      control: { type: 'radio' },
    },
  },
};
export default Story;

const Template: ComponentStory<typeof BaseTextBase> = (args: BaseTextProps) => (
  <StorybookScreen>
    <BaseTextBase {...args} />
  </StorybookScreen>
);

const parameters = {
  controls: {
    include: ['variant', 'weight', 'children'],
  },
};

const defaultArgs: BaseTextProps = {
  variant: TextVariant.H1,
  weight: TextWeight.Bold,
  children: 'This is BaseText component',
};

export const BaseText = Template.bind({});

BaseText.parameters = parameters;
BaseText.args = {
  ...defaultArgs,
};
