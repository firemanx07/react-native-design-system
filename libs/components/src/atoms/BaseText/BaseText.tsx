import {
  FontFamilyType,
  styled,
  useNamespacedTheme,
  FontSizeType,
  LineHeightType,
} from '@proxym/themes';
import React, { memo, useMemo } from 'react';
import { StyleProp, TextProps, TextStyle } from 'react-native';

export enum TextVariant {
  H1 = 'H1',
  H2 = 'H2',
  H3 = 'H3',
  L = 'L',
  M = 'M',
  S = 'S',
  XS = 'XS',
}

export enum TextWeight {
  Regular = 'Regular',
  SemiBold = 'SemiBold',
  Bold = 'Bold',
  ExtraBold = 'ExtraBold',
}

export type PropsType = {
  variant?: TextVariant;
  weight?: TextWeight;
  style?: StyleProp<TextStyle>;
} & TextProps;

export const BaseText = ({
  children,
  variant = TextVariant.L,
  weight = TextWeight.Regular,
  style,
  ...textProps
}: PropsType) => {
  const { fonts } = useNamespacedTheme();

  const fontFamily = useMemo(() => {
    switch (weight) {
      case TextWeight.Bold:
        return fonts.family.primaryBold;
      case TextWeight.SemiBold:
        return fonts.family.primarySemiBold;
      case TextWeight.ExtraBold:
        return fonts.family.primaryExtraBold;
      case TextWeight.Regular:
      default:
        return fonts.family.primary;
    }
  }, [weight]);

  const textStyles = useMemo(() => {
    switch (variant) {
      case TextVariant.H1:
        return {
          fontFamily: fonts.family.primaryBold,
          fontSize: fonts.size.large,
          lineHeight: fonts.lineHeight.medium,
        };
      case TextVariant.H2:
        return {
          fontFamily: fonts.family.primaryBold,
          fontSize: fonts.size.mediumLarge,
          lineHeight: fonts.lineHeight.mediumSmall,
        };
      case TextVariant.H3:
        return {
          fontFamily: fonts.family.primaryBold,
          fontSize: fonts.size.medium,
          lineHeight: fonts.lineHeight.mediumTiny,
        };
      case TextVariant.M:
        return {
          fontSize: fonts.size.small,
          lineHeight: fonts.lineHeight.mediumTiny,
        };
      case TextVariant.S:
        return {
          fontSize: fonts.size.extraSmall,
          lineHeight: fonts.lineHeight.small,
        };
      case TextVariant.XS:
        return {
          fontSize: fonts.size.tiny,
          lineHeight: fonts.lineHeight.extraSmall,
        };
      case TextVariant.L:
      default:
        return {
          fontSize: fonts.size.medium,
          lineHeight: fonts.lineHeight.mediumTiny,
        };
    }
  }, [variant]);

  return (
    <StyledText
      fontFamily={fontFamily}
      {...textStyles}
      {...textProps}
      style={style}
    >
      {children}
    </StyledText>
  );
};

type StyledTextVariant = {
  fontFamily: FontFamilyType;
  fontSize: FontSizeType;
  lineHeight: LineHeightType;
};

const StyledText = styled.Text<StyledTextVariant>`
  color: ${({ theme }) => theme.ds.colors.dark};
  font-family: ${({ fontFamily }) => fontFamily};
  font-size: ${({ fontSize }) => fontSize}px;
  line-height: ${({ lineHeight }) => lineHeight}px;
`;

export default memo(BaseText);
