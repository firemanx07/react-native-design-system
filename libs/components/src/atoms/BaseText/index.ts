export { default as BaseText, TextVariant, TextWeight } from './BaseText';
