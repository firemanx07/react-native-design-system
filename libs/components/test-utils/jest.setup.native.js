import { Animated } from 'react-native';

require('react-native-reanimated/lib/reanimated2/jestUtils').setUpTests();
require('react-native-gesture-handler/jestSetup.js');

console.warn = () => {
  //this is mock
};

Animated['timing'] = (value, config) => ({
  start: callback => {
    value.setValue(config.toValue);
    callback && callback({ finished: true });
  },
  stop: jest.fn(),
});
