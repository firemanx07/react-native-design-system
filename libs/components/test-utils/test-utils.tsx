import { theme, ThemeProvider } from '@celebrate-app/themes';
import { Story } from '@storybook/react';
import {
  cleanup,
  render,
  RenderOptions,
  waitFor,
} from '@testing-library/react-native';
import React, { ReactNode } from 'react';

const Providers = ({ children }: { children: ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions
) => {
  return render(ui, { wrapper: Providers, ...options });
};

beforeEach(cleanup);
afterEach(cleanup);

export function testStory(displayName: string, TestComponent: Story) {
  test(`${displayName} snapshot test`, async () => {
    const comp = customRender(<TestComponent {...TestComponent.args} />);
    jest.runOnlyPendingTimers();

    await waitFor(() => {
      expect(comp).toBeTruthy();
      expect(comp.toJSON()).toMatchSnapshot();
    });
  });
}

export function testComponentSnapshot(
  TestComponent: React.ReactElement<
    any,
    string | React.JSXElementConstructor<any>
  >
) {
  test('Snapshot test', () => {
    const comp = customRender(TestComponent);
    expect(comp).toBeTruthy();
    expect(comp.toJSON()).toMatchSnapshot();
  });
}

export function testAllStories(stories: any) {
  Object.keys(stories).forEach((story) => {
    if (story !== 'default') {
      testStory(story, stories[story]);
    }
  });
}
