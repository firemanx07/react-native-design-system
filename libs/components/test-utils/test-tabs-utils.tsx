import { waitFor } from '@testing-library/react-native';

import { customRender } from './test-utils';

type renderTabsProps = {
  elem: JSX.Element;
  testIDs: string[];
};
export const renderTabs = ({ elem, testIDs }: renderTabsProps) => {
  const comp = customRender(elem);
  const tabs: ReturnType<typeof comp.getByTestId>[] = [];
  testIDs.forEach(testID => tabs.push(comp.getByTestId(testID)));
  return { comp, tabs };
};
export const testTabs = async (args: ReturnType<typeof renderTabs>) => {
  await waitFor(() => {
    expect(args.comp).toBeTruthy();
    args.tabs.forEach(tab => expect(tab).toBeDefined());
  });
};
