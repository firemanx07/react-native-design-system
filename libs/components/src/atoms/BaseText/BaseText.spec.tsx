import { testAllStories } from '../../../test-utils';
import * as stories from './BaseText.stories';

describe('BaseText tests', () => {
  testAllStories(stories);
});
