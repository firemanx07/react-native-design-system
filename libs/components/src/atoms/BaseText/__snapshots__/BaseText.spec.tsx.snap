// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseText tests BaseText snapshot test 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "minHeight": 600,
        "paddingBottom": 16,
        "paddingLeft": 16,
        "paddingRight": 16,
        "paddingTop": 16,
      },
    ]
  }
>
  <Text
    fontFamily="Rubik-Bold"
    fontSize={20}
    lineHeight={26}
    style={
      [
        {
          "color": "#2C2C2C",
          "fontFamily": "Rubik-Bold",
          "fontSize": 20,
          "lineHeight": 26,
        },
      ]
    }
  >
    This is BaseText component
  </Text>
</View>
`;
